### PR TITLE
Bjorncs/improve athenz access control

### PR DESCRIPTION
**Pull request by bjorncs**
Requires https://github.com/vespa-engine/vespa/pull/13930 to be merged.

FYI @gjoranv @tokle

### CODE REVIEW

File: config-model/src/main/java/com/yahoo/vespa/model/container/http/AccessControl.java [hunk 1]
```diff
@@ -137,6 +139,22 @@ private void addAccessControlExcludedChain(Http http) {
         }
     }
 
+    // Remove bindings from access control chain that have binding pattern as a different filter chain
+    private void removeDuplicateBindingsFromAccessControlChain(Http http) {
+        Set<FilterBinding> duplicateBindings = new HashSet<>();
+        for (FilterBinding binding : http.getBindings()) {
+            if (binding.filterId().toId().equals(ACCESS_CONTROL_CHAIN_ID)) {
```
Comment (gjoranv) [inline config-model/src/main/java/com/yahoo/vespa/model/container/http/AccessControl.java:146]: BTW, is `filterId()` the correct name? Sounds like it should be `chainId`.

**Review by gjoranv — COMMENTED** [review]
It's great that you removed the code duplication in the builders. Otherwise, please see comments.

**Comment by gjoranv** [comment]
Build fails with:
```
java.lang.IllegalArgumentException: Could not create a component with id 'com.yahoo.jdisc.http.filter.security.misc.NoopFilter'. Tried to load class directly, since no bundle was found for spec: jdisc-security-filters. If a bundle with the same name is installed, there is a either a version mismatch or the installed bundle's version contains a qualifier string.
```

**Comment by bjorncs** [comment]
> Build fails with:
> 
> ```
> java.lang.IllegalArgumentException: Could not create a component with id 'com.yahoo.jdisc.http.filter.security.misc.NoopFilter'. Tried to load class directly, since no bundle was found for spec: jdisc-security-filters. If a bundle with the same name is installed, there is a either a version mismatch or the installed bundle's version contains a qualifier string.
> ```

This is fixed with https://github.com/vespa-engine/vespa/pull/13942/commits/8373e8d51ff3019e94f82211df16efae23080fe6

File: config-model/src/main/java/com/yahoo/vespa/model/container/http/AccessControl.java [hunk 2]
```diff
@@ -84,69 +85,75 @@ public AccessControl build() {
...
         this.writeEnabled = writeEnabled;
         this.excludedBindings = Collections.unmodifiableSet(excludedBindings);
         this.handlers = handlers;
-        this.servlets = servlets;
     }
 
-    public List<FilterBinding> getBindings() {
-        return Stream.concat(getHandlerBindings(), getServletBindings())
-                .collect(Collectors.toCollection(ArrayList::new));
+    public void configure(Http http) {
```
[the remark this answers is not in the document]
Comment (bjorncs) [reply config-model/src/main/java/com/yahoo/vespa/model/container/http/AccessControl.java:99]: Perhaps rename it to something like `configureAccessControlFilterChains`?

I would like to contain the logic for setting up filter chains and bindings inside the `AccessControl` class.

File: config-model/src/main/java/com/yahoo/vespa/model/container/component/SystemBindingPattern.java [hunk 3]
```diff
@@ -13,6 +13,7 @@
 
     public static SystemBindingPattern fromHttpPath(String path) { return new SystemBindingPattern("http", "*", null, path);}
     public static SystemBindingPattern fromPattern(String binding) { return new SystemBindingPattern(binding);}
+    public static SystemBindingPattern fromPortAndPath(String port, String path) { return new SystemBindingPattern("http", "*", port, path); }
```
[the remark this answers is not in the document]
Comment (bjorncs) [reply config-model/src/main/java/com/yahoo/vespa/model/container/component/SystemBindingPattern.java:16]: I agree, let's rename it to `fromHttpPortAndPath`.

File: config-model/src/main/java/com/yahoo/vespa/model/container/http/AccessControl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (gjoranv) [reply config-model/src/main/java/com/yahoo/vespa/model/container/http/AccessControl.java:99]: Isn't `configureAccessControl` enough? I'm ok with with either. I agree with containing the logic.

File: config-model/src/main/java/com/yahoo/vespa/model/container/http/AccessControl.java [hunk 4]
```diff
@@ -137,6 +139,22 @@ private void addAccessControlExcludedChain(Http http) {
         }
     }
 
+    // Remove bindings from access control chain that have binding pattern as a different filter chain
+    private void removeDuplicateBindingsFromAccessControlChain(Http http) {
+        Set<FilterBinding> duplicateBindings = new HashSet<>();
+        for (FilterBinding binding : http.getBindings()) {
```
[the remark this answers is not in the document]
Comment (bjorncs) [reply config-model/src/main/java/com/yahoo/vespa/model/container/http/AccessControl.java:133]: There are multiple strategies to solve this specific problem, but I don't think creating a mapping will be significantly better.

File: config-model/src/test/java/com/yahoo/vespa/model/container/xml/AccessControlTest.java [hunk 5]
```diff
@@ -44,15 +42,14 @@
 
     @Test
     public void access_control_filter_chains_are_set_up() {
-        Element clusterElem = DomBuilderTest.parse(
+        Http http = createModelAndGetHttp(
+                "<container version='1.0'>",
```
[the remark this answers is not in the document]
Comment (bjorncs) [reply config-model/src/test/java/com/yahoo/vespa/model/container/xml/AccessControlTest.java]: Fixed

File: config-model/src/main/java/com/yahoo/vespa/model/container/http/AccessControl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (bjorncs) [reply config-model/src/main/java/com/yahoo/vespa/model/container/http/AccessControl.java]: Renamed to `configureHttpFilterChains`.

**Comment by bjorncs** [comment]
@gjoranv PTAL

**Review by gjoranv — APPROVED** [review]
👍

**Pull request merged**